src/core/geometry/qgscircularstring.cpp

```diff
@@ -1592,6 +1594,118 @@ double QgsCircularString::segmentLength( QgsVertexId startVertex ) const
   return QgsGeometryUtilsBase::circleLength( x1, y1, x2, y2, x3, y3 );
 }
 
+double QgsCircularString::distanceBetweenVertices( QgsVertexId fromVertex, QgsVertexId toVertex ) const
+{
+  // Convert QgsVertexId to simple vertex numbers for curves (single ring, single part)
+  if ( fromVertex.part != 0 || fromVertex.ring != 0 || toVertex.part != 0 || toVertex.ring != 0 )
+    return -1.0;
+
+  int fromVertexNumber = fromVertex.vertex;
+  int toVertexNumber = toVertex.vertex;
+
+  if ( fromVertexNumber < 0 || fromVertexNumber >= numPoints() || toVertexNumber < 0 || toVertexNumber >= numPoints() )
+    return -1.0;
+
+  if ( fromVertexNumber == toVertexNumber )
+    return 0.0;
+
+  // Ensure fromVertex < toVertex for simplicity
+  if ( fromVertexNumber > toVertexNumber )
+  {
+    return distanceBetweenVertices( QgsVertexId( 0, 0, toVertexNumber ), QgsVertexId( 0, 0, fromVertexNumber ) );
+  }
+
+  double totalDistance = 0.0;
+  int nPoints = numPoints();
+
+  // Iterate through the arcs, accumulating distance between fromVertex and toVertex
+  for ( int i = 0; i < nPoints - 2; i += 2 )
+  {
+    // Arc segment from i to i+2, with curve point at i+1
+    double x1 = mX.at( i );     // Start point
+    double y1 = mY.at( i );
+    double x2 = mX.at( i + 1 ); // Curve point
+    double y2 = mY.at( i + 1 );
+    double x3 = mX.at( i + 2 ); // End point
+    double y3 = mY.at( i + 2 );
+
+    // Check if both vertices are in this arc segment
+    if ( fromVertexNumber >= i && toVertexNumber <= i + 2 )
+    {
+      if ( fromVertexNumber == i && toVertexNumber == i + 2 )
+      {
+        // Full arc from start to end
+        return QgsGeometryUtilsBase::circleLength( x1, y1, x2, y2, x3, y3 );
+      }
+      else if ( fromVertexNumber == i && toVertexNumber == i + 1 )
+      {
+        // Arc from start point to curve point - use precise calculation
+        double centerX, centerY, radius;
+        QgsGeometryUtils::circleCenterRadius( QgsPoint( x1, y1 ), QgsPoint( x2, y2 ), QgsPoint( x3, y3 ), radius, centerX, centerY );
```

Comment:
Shouldn't this use QgsGeometryUtilsBase::circleCenterRadius to avoid the cost of QgsPoint construction?

Reply:
Indeed